README.md
# gtirb-test-helpers

Provides helper functions to make creating GTIRB IR in tests easier and more
concise.

## Code generation

The gtirb-generate-python-code package can generate gtirb-test-helpers code
from a chunk of assembly code. This can serve as a good starting point to
generating the specific GTIRB that a test needs.

## Copyright and Acknowledgments

Copyright (C) 2021 GrammaTech, Inc.

This code is licensed under the MIT license. See the LICENSE file in
the project root for license terms.

This project is sponsored by the Office of Naval Research, One Liberty
Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
N68335-17-C-0700.  The content of the information does not necessarily
reflect the position or policy of the Government and no official
endorsement should be inferred.


CHANGES.txt
2.0.1 (Unreleased):

2.0.0:
- Removed additional_entries from add_function
- Added support for Python typing
- Exposed add_section to __init__ so it can be imported

1.3.0:
- add_code_block and add_data_block now accept (int, int) tuples as keys for
  symbolic expressions, with the first item being the offset and the second
  item being the size (which gets put in the symbolicExpressionSizes aux data
  table)
- add_function now accepts a set of entry blocks for the entry_block parameter

1.2.0:
- Added additional_entries parameter to add_function for specifying multiple
  entry points to functions

1.1.0:
- Added byte_order to create_test_module and defaulted to inferring it from
  the module's ISA

1.0.0:
- Initial version


tests/test_helpers.py
#
# Copyright (C) 2021 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#

import gtirb
from gtirb_test_helpers import (
    add_code_block,
    add_data_block,
    add_data_section,
    add_edge,
    add_elf_symbol_info,
    add_function,
    add_proxy_block,
    add_section,
    add_symbol,
    add_text_section,
    create_test_module,
    set_all_blocks_alignment,
)


def test_create_module():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.PE, gtirb.Module.ISA.ARM
    )
    assert ir
    assert m.ir is ir
    assert m.name
    assert m.isa == gtirb.Module.ISA.ARM
    assert m.file_format == gtirb.Module.FileFormat.PE
    assert m.byte_order == gtirb.Module.ByteOrder.Little
    assert m.aux_data


def test_create_module_with_binary_type():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.ELF,
        gtirb.Module.ISA.X64,
        binary_type=("DYN",),
        byte_order=gtirb.Module.ByteOrder.Big,
    )
    assert ir
    assert m.ir is ir
    assert m.name
    assert m.isa == gtirb.Module.ISA.X64
    assert m.file_format == gtirb.Module.FileFormat.ELF
    assert m.byte_order == gtirb.Module.ByteOrder.Big
    assert m.aux_data
    assert m.aux_data["binaryType"].data == ["DYN"]


def test_add_text_section():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    s, bi = add_text_section(m)
    assert s.name == ".text"
    assert bi.section is s
    assert bi.size == 0
    assert bi.contents == b""
    assert not bi.address


def test_add_text_section_with_address():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    s, bi = add_text_section(m, address=0x1000)
    assert s.name == ".text"
    assert bi.section is s
    assert bi.size == 0
    assert bi.contents == b""
    assert bi.address == 0x1000


def test_add_data_section():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    s, bi = add_data_section(m)
    assert s.name == ".data"
    assert bi.section is s
    assert bi.size == 0
    assert bi.contents == b""
    assert not bi.address


def test_add_data_section_with_address():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    s, bi = add_data_section(m, address=0x1000)
    assert s.name == ".data"
    assert bi.section is s
    assert bi.size == 0
    assert bi.contents == b""
    assert bi.address == 0x1000


def test_add_code_block():
    _, m = create_test_module(
        isa=gtirb.Module.ISA.X64, file_format=gtirb.Module.FileFormat.ELF
    )
    _, bi = add_text_section(m)
    b1 = add_code_block(bi, b"\xC3")
    assert bi.size == 1
    assert bi.contents == b"\xC3"
    assert not bi.symbolic_expressions
    assert isinstance(b1, gtirb.CodeBlock)
    assert b1.byte_interval is bi
    assert b1.offset == 0
    assert b1.size == 1

    expr1 = gtirb.SymAddrConst(0, gtirb.Symbol("hi"))
    expr2 = gtirb.SymAddrConst(0, gtirb.Symbol("bye"))
    b2 = add_code_block(
        bi, b"\x90\x90\xE8\x00\x00\x00\x00", {(0, 2): expr1, 3: expr2}
    )
    assert bi.size == 8
    assert bi.contents == b"\xC3\x90\x90\xE8\x00\x00\x00\x00"
    assert bi.symbolic_expressions == {1: expr1, 4: expr2}
    assert m.aux_data["symbolicExpressionSizes"].data == {
        gtirb.Offset(bi, 1): 2
    }
    assert isinstance(b2, gtirb.CodeBlock)
    assert b2.byte_interval is bi
    assert b2.offset == 1
    assert b2.size == 7


def test_add_data_block():
    _, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    _, bi = add_data_section(m)
    b1 = add_data_block(bi, b"\xC3")
    assert bi.size == 1
    assert bi.contents == b"\xC3"
    assert not bi.symbolic_expressions
    assert isinstance(b1, gtirb.DataBlock)
    assert b1.byte_interval is bi
    assert b1.offset == 0
    assert b1.size == 1

    expr = gtirb.SymAddrConst(0, gtirb.Symbol("hi"))
    b2 = add_data_block(bi, b"\x90\x90\xE8\x00\x00\x00\x00", {3: expr})
    assert bi.size == 8
    assert bi.contents == b"\xC3\x90\x90\xE8\x00\x00\x00\x00"
    assert bi.symbolic_expressions == {4: expr}
    assert isinstance(b2, gtirb.DataBlock)
    assert b2.byte_interval is bi
    assert b2.offset == 1
    assert b2.size == 7


def test_add_proxy_block():
    _, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    b1 = add_proxy_block(m)
    assert b1.module is m

    b2 = add_proxy_block(m)
    assert b1 is not b2
    assert b2.module is m


def test_add_symbol():
    _, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    proxy = add_proxy_block(m)
    sym = add_symbol(m, "hi", proxy)
    assert sym.module == m
    assert sym.name == "hi"
    assert sym.referent is proxy


def test_add_elf_symbol_info():
    _, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    sym = add_symbol(m, "hi")
    add_elf_symbol_info(m, sym, 100, "FUNC", "GLOBAL", "DEFAULT", 1)
    assert m.aux_data["elfSymbolInfo"].data == {
        sym: (100, "FUNC", "GLOBAL", "DEFAULT", 1)
    }


def test_add_function_with_name():
    _, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    _, bi = add_text_section(m)
    b1 = add_code_block(bi, b"\x90")
    b2 = add_code_block(bi, b"\xC3")
    b3 = add_code_block(bi, b"\xCC")

    func_id = add_function(m, "hello", {b1, b2}, {b3})
    assert func_id
    (sym,) = m.symbols
    assert sym.name == "hello"
    assert sym.referent is b1
    assert m.aux_data["functionEntries"].data[func_id] == {b1, b2}
    assert m.aux_data["functionBlocks"].data[func_id] == {b1, b2, b3}
    assert m.aux_data["functionNames"].data[func_id] == sym
    assert m.aux_data["elfSymbolInfo"].data[sym] == (
        0,
        "FUNC",
        "GLOBAL",
        "DEFAULT",
        0,
    )


def test_add_function_with_symbol():
    _, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    _, bi = add_text_section(m)
    b1 = add_code_block(bi, b"\x90")
    b2 = add_code_block(bi, b"\xC3")
    b3 = add_code_block(bi, b"\xCC")
    sym = add_symbol(m, "hello", b1)

    func_id = add_function(m, sym, b1, {b2, b3})
    assert func_id
    assert m.aux_data["functionEntries"].data[func_id] == {b1}
    assert m.aux_data["functionBlocks"].data[func_id] == {b1, b2, b3}
    assert m.aux_data["functionNames"].data[func_id] == sym
    assert m.aux_data["elfSymbolInfo"].data[sym] == (
        0,
        "FUNC",
        "GLOBAL",
        "DEFAULT",
        0,
    )


def test_add_edge():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    _, bi = add_text_section(m)
    b1 = add_code_block(bi, b"\x90")
    b2 = add_code_block(bi, b"\xC3")

    edge = add_edge(
        ir.cfg, b1, b2, gtirb.Edge.Type.Branch, conditional=True, direct=False
    )
    assert edge in ir.cfg
    assert edge.source is b1
    assert edge.target is b2
    assert edge.label
    assert edge.label.type == gtirb.Edge.Type.Branch
    assert edge.label.conditional
    assert not edge.label.direct


def test_set_all_blocks_alignment():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    _, bi = add_text_section(m)
    b1 = add_code_block(bi, b"\x90")
    b2 = add_code_block(bi, b"\xC3")
    b3 = add_code_block(bi, b"\xCC")

    set_all_blocks_alignment(m, 1)
    assert m.aux_data["alignment"].data == {b1: 1, b2: 1, b3: 1}


def test_add_section():
    ir, m = create_test_module(
        gtirb.Module.FileFormat.ELF, gtirb.Module.ISA.X64
    )
    s1, bi1 = add_section(m, ".init_array")

    assert s1.name == ".init_array"
    assert s1.byte_intervals == {bi1}
    assert s1.flags == {
        gtirb.Section.Flag.Readable,
        gtirb.Section.Flag.Loaded,
        gtirb.Section.Flag.Initialized,
    }
    assert bi1.contents == b""
    assert bi1.address is None

    s2, bi2 = add_section(m, ".fini_array", 0x1000, flags=set())
    assert s2.name == ".fini_array"
    assert s2.byte_intervals == {bi2}
    assert s2.flags == set()
    assert bi2.contents == b""
    assert bi2.address == 0x1000


gtirb_test_helpers/__init__.py
#
# Copyright (C) 2021 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#

from .helpers import (
    add_code_block,
    add_data_block,
    add_data_section,
    add_edge,
    add_elf_symbol_info,
    add_function,
    add_proxy_block,
    add_symbol,
    add_text_section,
    create_test_module,
    set_all_blocks_alignment,
)
from .version import __version__

__all__ = [
    "add_data_section",
    "add_edge",
    "add_code_block",
    "add_data_block",
    "add_elf_symbol_info",
    "add_function",
    "add_proxy_block",
    "add_symbol",
    "add_text_section",
    "create_test_module",
    "set_all_blocks_alignment",
    "__version__",
]


gtirb_test_helpers/helpers.py
#
# Copyright (C) 2021 GrammaTech, Inc.
#
# This code is licensed under the MIT license. See the LICENSE file in
# the project root for license terms.
#
# This project is sponsored by the Office of Naval Research, One Liberty
# Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
# N68335-17-C-0700.  The content of the information does not necessarily
# reflect the position or policy of the Government and no official
# endorsement should be inferred.
#

"""
Utilities for constructing GTIRB IR in tests.
"""

import uuid
from typing import Dict, Iterable, Set, Tuple, Type, TypeVar, Union

import gtirb

_DEFAULT_BYTE_ORDERS = {
    gtirb.Module.ISA.ARM: gtirb.Module.ByteOrder.Little,
    gtirb.Module.ISA.ARM64: gtirb.Module.ByteOrder.Little,
    gtirb.Module.ISA.IA32: gtirb.Module.ByteOrder.Little,
    gtirb.Module.ISA.PPC32: gtirb.Module.ByteOrder.Big,
    gtirb.Module.ISA.PPC64: gtirb.Module.ByteOrder.Big,
    gtirb.Module.ISA.X64: gtirb.Module.ByteOrder.Little,
}


def create_test_module(
    file_format: gtirb.Module.FileFormat,
    isa: gtirb.Module.ISA,
    binary_type: Iterable[str] = None,
    byte_order: gtirb.Module.ByteOrder = None,
) -> Tuple[gtirb.IR, gtirb.Module]:
    """
    Creates a test GTIRB module and returns the IR object and module object.
    """

    if byte_order is None:
        actual_byte_order = _DEFAULT_BYTE_ORDERS.get(
            isa, gtirb.Module.ByteOrder.Undefined
        )
    else:
        actual_byte_order = byte_order

    ir = gtirb.IR()
    m = gtirb.Module(
        isa=isa,
        file_format=file_format,
        name="test",
        byte_order=actual_byte_order,
    )
    m.ir = ir

    add_standard_aux_data_tables(m)
    if binary_type:
        m.aux_data["binaryType"].data = list(binary_type)

    return ir, m


def add_standard_aux_data_tables(m: gtirb.Module) -> None:
    """
    Adds aux data tables that ddisasm normally produces. This helps avoid
    every test needing to know the schemas.
    """
    m.aux_data["binaryType"] = gtirb.AuxData(
        type_name="sequence<string>", data=list()
    )
    m.aux_data["cfiDirectives"] = gtirb.AuxData(
        type_name=(
            "mapping<Offset,sequence<tuple<string,sequence<int64_t>,UUID>>>"
        ),
        data=dict(),
    )
    m.aux_data["comments"] = gtirb.AuxData(
        type_name="mapping<Offset,string>", data=dict()
    )
    m.aux_data["elfSectionProperties"] = gtirb.AuxData(
        type_name="mapping<UUID,tuple<uint64_t,uint64_t>>", data=dict()
    )
    m.aux_data["encodings"] = gtirb.AuxData(
        type_name="mapping<UUID,string>", data=dict()
    )
    m.aux_data["functionBlocks"] = gtirb.AuxData(
        type_name="mapping<UUID,set<UUID>>", data=dict()
    )
    m.aux_data["functionEntries"] = gtirb.AuxData(
        type_name="mapping<UUID,set<UUID>>", data=dict()
    )
    m.aux_data["functionNames"] = gtirb.AuxData(
        type_name="mapping<UUID,UUID>", data=dict()
    )
    m.aux_data["libraries"] = gtirb.AuxData(
        type_name="sequence<string>", data=list()
    )
    m.aux_data["libraryPaths"] = gtirb.AuxData(
        type_name="sequence<string>", data=list()
    )
    m.aux_data["padding"] = gtirb.AuxData(
        type_name="mapping<Offset,uint64_t>", data=dict()
    )
    m.aux_data["symbolForwarding"] = gtirb.AuxData(
        type_name="mapping<UUID,UUID>", data=dict()
    )
    m.aux_data["symbolicExpressionSizes"] = gtirb.AuxData(
        type_name="mapping<Offset,uint64_t>", data=dict()
    )
    m.aux_data["SCCs"] = gtirb.AuxData(
        type_name="mapping<UUID,int64_t>", data=dict()
    )

    if m.file_format == gtirb.Module.FileFormat.ELF:
        m.aux_data["alignment"] = gtirb.AuxData(
            type_name="mapping<UUID,uint64_t>", data=dict()
        )
        m.aux_data["dynamicEntries"] = gtirb.AuxData(
            type_name="set<tuple<string,uint64_t>>", data=set()
        )
        m.aux_data["elfSymbolInfo"] = gtirb.AuxData(
            type_name=(
                "mapping<UUID,tuple<uint64_t,string,string,string,uint64_t>>"
            ),
            data=dict(),
        )
        m.aux_data["elfSymbolTabIdxInfo"] = gtirb.AuxData(
            type_name="mapping<UUID,sequence<tuple<string,uint64_t>>>",
            data=dict(),
        )

    elif m.file_format == gtirb.Module.FileFormat.PE:
        m.aux_data["peExportEntries"] = gtirb.AuxData(
            type_name="sequence<tuple<uint64_t,int64_t,string>>", data=list()
        )
        m.aux_data["peExportedSymbols"] = gtirb.AuxData(
            type_name="sequence<UUID>", data=list()
        )
        m.aux_data["peImportEntries"] = gtirb.AuxData(
            type_name="sequence<tuple<uint64_t,int64_t,string,string>>",
            data=list(),
        )
        m.aux_data["peImportedSymbols"] = gtirb.AuxData(
            type_name="sequence<UUID>", data=list()
        )
        m.aux_data["peResources"] = gtirb.AuxData(
            type_name="sequence<tuple<sequence<uint8_t>,Offset,uint64_t>>",
            data=list(),
        )


def add_section(
    m: gtirb.Module,
    name: str,
    address: int = None,
    flags: Set[gtirb.Section.Flag] = {
        gtirb.Section.Flag.Readable,
        gtirb.Section.Flag.Loaded,
        gtirb.Section.Flag.Initialized,
    },
) -> Tuple[gtirb.Section, gtirb.ByteInterval]:
    """
    Adds a section to the module and creates a byte interval for it.
    """
    s = gtirb.Section(name=name, flags=flags)
    s.module = m
    bi = gtirb.ByteInterval(contents=b"", address=address)
    bi.section = s
    return s, bi


def add_text_section(
    m: gtirb.Module, address: int = None
) -> Tuple[gtirb.Section, gtirb.ByteInterval]:
    """
    Creates the appropriate text section for a module.
    """
    assert m.file_format in (
        gtirb.Module.FileFormat.ELF,
        gtirb.Module.FileFormat.PE,
    )
    return add_section(
        m,
        ".text",
        address,
        flags={
            gtirb.Section.Flag.Readable,
            gtirb.Section.Flag.Executable,
            gtirb.Section.Flag.Loaded,
            gtirb.Section.Flag.Initialized,
        },
    )


def add_data_section(
    m: gtirb.Module, address: int = None
) -> Tuple[gtirb.Section, gtirb.ByteInterval]:
    """
    Creates the appropriate data section for a module.
    """
    assert m.file_format in (
        gtirb.Module.FileFormat.ELF,
        gtirb.Module.FileFormat.PE,
    )
    return add_section(
        m,
        ".data",
        address,
        flags={
            gtirb.Section.Flag.Readable,
            gtirb.Section.Flag.Writable,
            gtirb.Section.Flag.Loaded,
            gtirb.Section.Flag.Initialized,
        },
    )


BlockT = TypeVar("BlockT", bound=gtirb.ByteBlock)


def add_byte_block(
    byte_interval: gtirb.ByteInterval,
    block_type: Type[BlockT],
    content: bytes,
    symbolic_expressions: Dict[
        Union[int, Tuple[int, int]], gtirb.SymbolicExpression
    ] = None,
) -> BlockT:
    """
    Adds a block to a byte interval, setting up its contents and optionally
    its symbolic expressions.
    :param byte_interval: The byte interval to add the block to.
    :param block_type: The block class to use.
    :param contents: The content of the block.
    :param symbolic_expressions: A dict mapping either symbolic expression
           offset or tuples of [symbolic expression offset, symbolic
           expression size] to symbplic expression.
    :returns: The new block.
    """
    b = block_type(offset=byte_interval.size, size=len(content))
    b.byte_interval = byte_interval
    byte_interval.contents += content

    if (
        byte_interval.module
        and "symbolicExpressionSizes" in byte_interval.module.aux_data
    ):
        size_aux_data = byte_interval.module.aux_data[
            "symbolicExpressionSizes"
        ].data
    else:
        size_aux_data = None
        for key in symbolic_expressions.keys():
            if not isinstance(key, int):
                raise ValueError("could not find symbolicExpressionSizes")

    if symbolic_expressions:
        for key, expr in symbolic_expressions.items():
            assert isinstance(expr, gtirb.SymbolicExpression)
            if isinstance(key, int):
                off = key
            else:
                off, size = key
                size_aux_data[
                    gtirb.Offset(byte_interval, byte_interval.size + off)
                ] = size

            byte_interval.symbolic_expressions[byte_interval.size + off] = expr
    byte_interval.size += len(content)
    return b


def add_code_block(
    byte_interval: gtirb.ByteInterval,
    content: bytes,
    symbolic_expressions: Dict[
        Union[int, Tuple[int, int]], gtirb.SymbolicExpression
    ] = None,
) -> gtirb.CodeBlock:
    """
    Adds a code block to a byte interval, setting up its contents and
    optionally its symbolic expressions.
    :param byte_interval: The byte interval to add the block to.
    :param contents: The content of the block.
    :param symbolic_expressions: A dict mapping either symbolic expression
           offset or tuples of [symbolic expression offset, symbolic
           expression size] to symbplic expression.
    :returns: The new block.
    """
    return add_byte_block(
        byte_interval, gtirb.CodeBlock, content, symbolic_expressions
    )


def add_data_block(
    byte_interval: gtirb.ByteInterval,
    content: bytes,
    symbolic_expressions: Dict[
        Union[int, Tuple[int, int]], gtirb.SymbolicExpression
    ] = None,
) -> gtirb.DataBlock:
    """
    Adds a data block to a byte interval, setting up its contents and
    optionally its symbolic expressions.
    :param byte_interval: The byte interval to add the block to.
    :param contents: The content of the block.
    :param symbolic_expressions: A dict mapping either symbolic expression
           offset or tuples of [symbolic expression offset, symbolic
           expression size] to symbplic expression.
    :returns: The new block.
    """
    return add_byte_block(
        byte_interval, gtirb.DataBlock, content, symbolic_expressions
    )


def add_proxy_block(m: gtirb.Module) -> gtirb.ProxyBlock:
    """
    Creates a proxy block and adds it to the module.
    """
    b = gtirb.ProxyBlock()
    m.proxies.add(b)
    return b


def add_symbol(
    module: gtirb.Module, name: str, payload: gtirb.Block = None
) -> gtirb.Symbol:
    """
    Creates a symbol and adds it to the module.
    """
    sym = gtirb.Symbol(name, payload=payload)
    sym.module = module
    return sym


def add_elf_symbol_info(
    module: gtirb.Module,
    sym: gtirb.Symbol,
    size: int,
    type: str,
    binding: str = "GLOBAL",
    visibility: str = "DEFAULT",
    section_index: int = 0,
) -> None:
    """
    Adds an entry to the elfSymbolInfo aux data table.
    """
    module.aux_data["elfSymbolInfo"].data[sym] = (
        size,
        type,
        binding,
        visibility,
        section_index,
    )


def add_function(
    module: gtirb.Module,
    sym_or_name: Union[str, gtirb.Symbol],
    entry_block: Union[gtirb.CodeBlock, Set[gtirb.CodeBlock]],
    other_blocks: Set[gtirb.CodeBlock] = set(),
) -> uuid.UUID:
    """
    Adds a function to all the appropriate aux data.
    :param module: The module to add the function to.
    :param sym_or_name: The function name, which can either be a string or a
                        symbol. If a string is passed, an appropriate symbol
                        is implicitly created. If there are multiple entry
                        blocks, that symbol will refer to the block with the
                        lowest offset.
    :param entry_block: The function's entry block or set of entry blocks.
    :param other_blocks: A set of other non-entry blocks in the function.
                         Optional.
    :returns: The new function's UUID.
    """

    if isinstance(entry_block, gtirb.CodeBlock):
        entry_blocks = {entry_block}
    else:
        entry_blocks = entry_block

    if isinstance(sym_or_name, str):
        best_entry = sorted(entry_blocks, key=lambda b: b.offset)[0]
        func_sym = add_symbol(module, sym_or_name, best_entry)
    elif isinstance(sym_or_name, gtirb.Symbol):
        func_sym = sym_or_name
    else:
        assert False, "Invalid symbol name"

    all_blocks = entry_blocks | other_blocks

    func_uuid = uuid.uuid4()
    module.aux_data["functionNames"].data[func_uuid] = func_sym
    module.aux_data["functionEntries"].data[func_uuid] = entry_blocks
    module.aux_data["functionBlocks"].data[func_uuid] = all_blocks

    if (
        module.file_format == gtirb.Module.FileFormat.ELF
        and "elfSymbolInfo" in module.aux_data
    ):
        add_elf_symbol_info(module, func_sym, 0, "FUNC")

    return func_uuid


def add_edge(
    cfg: gtirb.CFG,
    source: gtirb.CfgNode,
    target: gtirb.CfgNode,
    edge_type: gtirb.Edge.Type,
    *,
    conditional: bool = False,
    direct: bool = True,
) -> gtirb.Edge:
    """
    Creates and adds an edge to a CFG.
    """
    edge = gtirb.Edge(
        source=source,
        target=target,
        label=gtirb.Edge.Label(
            type=edge_type, conditional=conditional, direct=direct
        ),
    )
    cfg.add(edge)
    return edge


def set_all_blocks_alignment(module: gtirb.Module, alignment: int) -> None:
    """
    Sets the alignment of all blocks in a module to a specific value.
    """
    for block in module.byte_blocks:
        module.aux_data["alignment"].data[block] = alignment
